### PR TITLE
CEOI Harbingers

### PR DESCRIPTION
**Pull request by dutinmeow**
I even rewrote the code with documentation

### CODE REVIEW

File: solutions/ceoi-09-Harbingers.mdx [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+		T[u].emplace_back(v, w);
+		T[v].emplace_back(u, w);
+	}
+	for (int i = 2; i <= N; i++)
+		cin >> S[i] >> V[i];
+
+	dfs();
+
+	for (int i = 2; i <= N; i++)
+		cout << min_time[i] << ' ';
```
[the remark this answers is not in the document]
Comment (dutinmeow) [reply solutions/ceoi-09-Harbingers.mdx:154]: done

File: solutions/ceoi-09-Harbingers.mdx [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+ * into the convex hull, it would occupy that position. The 
+ * convex hull is stored in stk. If no lines are to be 
+ * removed to make room for the new line, then the size of the
+ * convex hull is returned. Note that the convex hull 
+ * is NOT modified in this function. 
+ * @param line The line to be queried
+ * @return An index between 1 and stk_max (inclusive) specifying 
+ *		 the location the given line would occupy should it 
+ *		 be added to the hull
+ */
```
Comment (Aadit-Ambadkar) [inline solutions/ceoi-09-Harbingers.mdx:89]:
```suggestion
```

Explain this in editorial if needed

File: solutions/ceoi-09-Harbingers.mdx [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	return l;
+}
+
+/**
+ * Finds the smallest dp value for all nodes in the 
+ * current subtree
+ * @param u Current node
+ * @param p Parent of current node
+ * @param d Distance from root
+ */
```
Comment (Aadit-Ambadkar) [inline solutions/ceoi-09-Harbingers.mdx:116]:
```suggestion
```

Explain this in editorial if needed

File: solutions/ceoi-09-Harbingers.mdx [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+Line stk[MAXN];  // convex hull of lines
+int stk_max = 0;  // end of convex hull stack
+
+/**
+ * Given the convex hull of points in stk[],
+ * finds the minimum y value for the given x value
+ * out of all lines in the hull
+ * @param x The x position
+ * @return The minimum y value in the convex hull
+ */
```
Comment (Aadit-Ambadkar) [inline solutions/ceoi-09-Harbingers.mdx:63]:
```suggestion
```

Explain this in editorial if needed

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: solutions/ceoi-09-Harbingers.mdx (hunk 2 above)
Comment (dutinmeow) [reply solutions/ceoi-09-Harbingers.mdx:89]: but all of this stuff is implementation specific?

File: solutions/ceoi-09-Harbingers.mdx (hunk 3 above)
Comment (dutinmeow) [reply solutions/ceoi-09-Harbingers.mdx:116]: ^^^

File: solutions/ceoi-09-Harbingers.mdx (hunk 4 above)
Comment (dutinmeow) [reply solutions/ceoi-09-Harbingers.mdx:63]: ^^^

File: solutions/ceoi-09-Harbingers.mdx (hunk 2 above)
Comment (Aadit-Ambadkar) [reply solutions/ceoi-09-Harbingers.mdx:89]: implementation-specific editorial. this much commenting is not needed

File: solutions/ceoi-09-Harbingers.mdx [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+		prev_line = stk[stk_max];  // store previous line at replacement position
+		stk[stk_max++] = l;  // update replacement position to new line
+	}
+
+	// recurse to children
+	for (auto [v, w] : T[u]) {
+		if (v != p) {
+			dfs(v, u, d + w);
+		}
+	}
```
Comment (Aadit-Ambadkar) [inline solutions/ceoi-09-Harbingers.mdx:137]:
```suggestion
	for (auto [v, w] : T[u]) if (v != p) dfs(v, u, d + w);
```
#codecompress

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: solutions/ceoi-09-Harbingers.mdx (hunk 2 above)
Comment (Aadit-Ambadkar) [reply solutions/ceoi-09-Harbingers.mdx:89]: implementation-specific editorial. this much commenting is not needed

File: solutions/ceoi-09-Harbingers.mdx (hunk 4 above)
Comment (SansPapyrus683) [reply solutions/ceoi-09-Harbingers.mdx:63]: what does this even mean

File: solutions/ceoi-09-Harbingers.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/ceoi-09-Harbingers.mdx:137]: NO. NONONONONONONONONONO.

File: solutions/ceoi-09-Harbingers.mdx (hunk 2 above)
Comment (dutinmeow) [reply solutions/ceoi-09-Harbingers.mdx:89]: It's better than nothing? how is more (useful) comments a bad thing? also, the editorial that is linked is pretty good, so there's no need to reiterate redundant information

File: solutions/ceoi-09-Harbingers.mdx (hunk 5 above)
Comment (dutinmeow) [reply solutions/ceoi-09-Harbingers.mdx:137]: but... this isn't consistant with the style guide?

**Comment by dutinmeow** [comment]
screw you stale bot

**Comment by SansPapyrus683** [comment]
dustin i don't see you making these changes

**Comment by dutinmeow** [comment]
I purposefully leave the explanation in the code, because it is either small descriptive comments or general implementation details. The algorithm itself is explained throughly in the editorial, and there's no need to reiterate information that's already publicly available (with the possible risk of plagiarism). The code is here to add extra depth.

File: solutions/ceoi-09-Harbingers.mdx [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+	cout << '\n';
+}
+```
+</CPPSection>
+
+</LanguageSection>
+
+<Info title="Alternative Solution">
+
+This problem can also be solved with a persistant [Li-Chao tree](https://cp-algorithms.com/geometry/convex_hull_trick.html#toc-tgt-1) (although the memory limit is rather tight).
```
Comment (SansPapyrus683) [inline solutions/ceoi-09-Harbingers.mdx:172]: "persistant"

File: solutions/ceoi-09-Harbingers.mdx [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+	Line(ll m = 0, ll b = 0) : m(m), b(b) {}
+
+	/** 
+	 * Evaluates the linear function at position x 
+	 */ 
+	ll operator()(ll x) { return m * x + b; }
+
+	/** 
+	 * Returns the x-coordinate of the intersection of two lines 
+	 */
```
Comment (SansPapyrus683) [inline solutions/ceoi-09-Harbingers.mdx:41]: make this doc comment one line like you did with the class doc comment

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Comment by SansPapyrus683** [comment]
@dutinmeow you said docs only show up w/ "/** */"
which ide are you using
image
image

**Comment by dutinmeow** [comment]
`/** */` is the way you're supposed to do documentation? It doesn't really matter that much

**Comment by SansPapyrus683** [comment]
yeah, but i was just curious

**Review by caoash — APPROVED** [review]

**Comment by SansPapyrus683** [comment]
two mods including me, that should be good enough

**Pull request merged**